src/algebra/add_comm_torsor.lean

```diff
@@ -0,0 +1,180 @@
+/-
+Copyright (c) 2020 Joseph Myers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Joseph Myers.
+-/
+import algebra.group.basic
+
+/-!
+# Torsors of additive commutative group actions
+
+This file defines torsors of additive commutative group actions.
+
+## Notations
+
+The group elements are referred to as vectors and the elements they
+act on are referred to as points, reflecting the use case of affine
+spaces.  This file defines the notation `+ᵥ` for adding a vector to a
+point and `-ᵥ` for subtracting two points to produce a vector.
+
+## Implementation notes
+
+It may be appropriate to refactor in future as a special case of
+torsors of additive group actions, in terms of the general definition
+of group actions (currently mathlib only has multiplicative group
+actions).  The variable `V` is an explicit rather than implicit
+argument to lemmas because otherwise the elaborator has problems
+inferring appropriate types and type class instances.
+
+## References
+
+* https://en.wikipedia.org/wiki/Principal_homogeneous_space
+* https://en.wikipedia.org/wiki/Affine_space
+
+-/
+
+/-- Type class for the `+ᵥ` and `-ᵥ` notation. -/
+class has_vadd (V : Type*) (P : Type*) :=
+(vadd : P → V → P)
+(vsub : P → P → V)
+
+infix `+ᵥ`:65 := has_vadd.vadd
+infix `-ᵥ`:65 := has_vadd.vsub
+
+section prio
+set_option default_priority 100 -- see Note [default priority]
+/-- An `add_comm_torsor V P` gives a structure to the nonempty type
+`P`, acted on by an `add_comm_group V` with a transitive and free
+action given by the `+ᵥ` operation and a corresponding subtraction
+given by the `-ᵥ` operation. In the case of a vector space, it is an
+affine space.  (The result of adding a zero vector does not need to be
+included here because it is deduced below from the other axioms.) -/
+class add_comm_torsor (V : Type*) (P : Type*) [add_comm_group V] [nonempty P]
+    extends has_vadd V P :=
+(vadd_assoc : ∀ (p : P) (v1 v2 : V), p +ᵥ v1 +ᵥ v2 = p +ᵥ (v1 + v2))
+(vadd_vsub : ∀ (p1 p2 : P), p1 +ᵥ (p2 -ᵥ p1 : V) = p2)
+(vsub_vadd : ∀ (p : P) (v : V), p +ᵥ v -ᵥ p = v)
+end prio
+
+/-- An `add_comm_group V` is a torsor for itself. -/
+instance add_comm_group_has_vadd (V : Type*) [add_comm_group V] : has_vadd V V :=
+{ vadd := has_add.add,
+  vsub := has_sub.sub }
+instance add_comm_group_is_add_comm_torsor (V : Type*) [nonempty V] [add_comm_group V] :
+  add_comm_torsor V V :=
+{ vadd_assoc := add_assoc,
+  vadd_vsub := λ a b, add_eq_of_eq_sub' rfl,
+  vsub_vadd := add_sub_cancel' }
+
+namespace add_comm_torsor
+
+variables (k : Type*) (V : Type*) {P : Type*} [add_comm_group V] [nonempty P]
+          [S : add_comm_torsor V P]
+include S
+
+/-- Adding two vectors to a point produces the same result as adding
+their sum. -/
+lemma vadd_add_assoc (p : P) (v1 v2 : V) : p +ᵥ v1 +ᵥ v2 = p +ᵥ (v1 + v2) :=
+add_comm_torsor.vadd_assoc p v1 v2
```

Comment:
since you are in the `add_comm_torsor` namespace, it should be enough to write `vadd_assoc p v1 v2`.

Reply:
If you are restating a version of a constructor of the class, but with different binders (here, `V` is explicit), the usual convention is to use the same name for both versions, with a prime in the constructor.

Reply:
I've made that change.